### ISSUE TITLE
Not compiling

### ISSUE THREAD

**Issue by LaserTron** [rbdrum2midi/rbdrum2midi#15]
Hi, I'm trying to compile this in Manjaro Linux. Following the instructions
```
mkdir build
cd build
cmake ..
make
```
produces the following:
```
[ 11%] Building C object src/CMakeFiles/rbdrum2midi.dir/jackdriver.c.o
[ 22%] Building C object src/CMakeFiles/rbdrum2midi.dir/main.c.o
[ 33%] Building C object src/CMakeFiles/rbdrum2midi.dir/alsadriver.c.o
[ 44%] Building C object src/CMakeFiles/rbdrum2midi.dir/ghkit.c.o
[ 55%] Building C object src/CMakeFiles/rbdrum2midi.dir/rb1kit.c.o
[ 66%] Building C object src/CMakeFiles/rbdrum2midi.dir/rbkit.c.o
[ 77%] Building C object src/CMakeFiles/rbdrum2midi.dir/rb3keyboard.c.o
[ 88%] Building C object src/CMakeFiles/rbdrum2midi.dir/rbguitar.c.o
[100%] Linking C executable rbdrum2midi
```
But produces the following errors
```
/usr/bin/ld: CMakeFiles/rbdrum2midi.dir/ghkit.c.o:/home/nick/rbdrum2midi/src/mididrum.h:170: multiple definition of `do_exit'; CMakeFiles/rbdrum2midi.dir/main.c.o:/home/nick/rbdrum2midi/src/mididrum.h:170: first defined here
/usr/bin/ld: CMakeFiles/rbdrum2midi.dir/rb1kit.c.o:/home/nick/rbdrum2midi/src/mididrum.h:170: multiple definition of `do_exit'; CMakeFiles/rbdrum2midi.dir/main.c.o:/home/nick/rbdrum2midi/src/mididrum.h:170: first defined here
/usr/bin/ld: CMakeFiles/rbdrum2midi.dir/rbkit.c.o:/home/nick/rbdrum2midi/src/mididrum.h:170: multiple definition of `do_exit'; CMakeFiles/rbdrum2midi.dir/main.c.o:/home/nick/rbdrum2midi/src/mididrum.h:170: first defined here
/usr/bin/ld: CMakeFiles/rbdrum2midi.dir/rb3keyboard.c.o:/home/nick/rbdrum2midi/src/mididrum.h:170: multiple definition of `do_exit'; CMakeFiles/rbdrum2midi.dir/main.c.o:/home/nick/rbdrum2midi/src/mididrum.h:170: first defined here
/usr/bin/ld: CMakeFiles/rbdrum2midi.dir/rbguitar.c.o:/home/nick/rbdrum2midi/src/mididrum.h:170: multiple definition of `do_exit'; CMakeFiles/rbdrum2midi.dir/main.c.o:/home/nick/rbdrum2midi/src/mididrum.h:170: first defined here
collect2: error: ld returned 1 exit status
make[2]: *** [src/CMakeFiles/rbdrum2midi.dir/build.make:212: src/rbdrum2midi] Error 1
make[1]: *** [CMakeFiles/Makefile2:98: src/CMakeFiles/rbdrum2midi.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```
There is no executable.

**Comment by thesteg** [comment]
Could be wrong but think the do_exit variable needs to be declared in the main.c file and marked extern in the mididrum.h file. 

Try just adding the line `int do_exit;` somewhere near the top of main.c, just under the #includes and popping an extern keyword at the start of the int do_exit line in the mididrum.h. (makes `extern int do_exit;`).

**Comment by randomas** [comment]
Hey there great to see people still interested in this! I fear the codebase is a little old Last time I tried compiling it was ubuntu 10.04, and possibly a test in 14.04. I fear compilers have become a little more finniky and the code base probably needs cleaning up accordingly. I'll give it a swirl as soon as I have access to a pc with a compiler on it. If you find a solution please let us know.

**Comment by thesteg** [comment]
Created a pull request, it now compiles but no way of knowing if it works as I no longer have the hardware!

**Comment by randomas** [comment]
> Created a pull request, it now compiles but no way of knowing if it works as I no longer have the hardware!

Fantastic, I still have the hardware, unfortunately it's in storage. I may get permission from the wife to make a mess in the house and pull it out, but not for the next few days at least. If the Op can test and let us know that would be great!

**Comment by LaserTron** [comment]
@thesteg, it compiles! I'll try it out tonight when I get back home.

**Comment by LaserTron** [comment]
Plugging in my game drum pad and running as root makes a midi thing show up, but I haven't been able to make any noise from it yet. I have a midi keyboard and that pretty much works out of the box.

![Screenshot from 2022-03-29 18-38-32](https://user-images.githubusercontent.com/12416002/160866739-b752e2e2-4ff6-4650-8e8d-bf42ef41ec8b.png)

**Comment by randomas** [comment]
Have you tried with hydrogen, that's what it was originally set up for. This is a picture from an old version on sourceforge, but should still be about the same.  You should connect the drumkit client to hydrogen and hydrogen to out.
![image](https://user-images.githubusercontent.com/10617922/160868963-909f622d-3c1d-442a-a5da-4b3e3a010923.png)

**Comment by LaserTron** [comment]
I actually did hook up the drum kit to hydrogen in Jack and it didn't work, that said, midi is finnicky and so are various application settings. What is the application you have in the terminal? Is it something that shows all the midi messages? I could probably use that to see if anything at all is being picked up from the drum kit (which I bought at a thrift store).

**Comment by ssj71** [comment]
There is a terminal application called `jack_midi_dump` that  creates a jack midi client and prints out all received messages.. Very handy.

**Comment by randomas** [comment]
The terminal in the picture is just showing rbdrum2midi running in debug mode. There should be a command line option otherwise there's a switch to uncomment in the opening lines in the main source file. I'd need to check.

**Comment by ssj71** [comment]
Just pass the arguments `-dbg -v` to make it verbose debug mode

**Comment by LaserTron** [comment]
Great news! Something is definitely being picked up! I think it's just a question of configuring my software to talk through midi properly.
![Screenshot from 2022-03-31 18-02-39](https://user-images.githubusercontent.com/12416002/161149045-c2d571df-9b3b-4aa8-a5fa-db4adf39a0d1.png)

**Comment by LaserTron** [comment]
Okay, so although `rbdrum2midi -dbg -v` shows events as I hit the pads I couldn't get any sounds in Hydrogen. I used
`amidi -p virtual -d` which creates a virtual midi port and writes to the terminal. (`jack_midi_dump` doesn't seem to be provided by my distro)

Events shown in the output of  `rbdrum2midi -dbg -v` don't show up in the output of `amidi -p virtual -d`. To test I also hooked up my midi keyboard. Key presses were shown in the output of `amidi -p virtual -d`. 

So it seems like the program recognizes and connects to the drum pad but it doesn't seem to send midi
![Screenshot from 2022-03-31 20-56-48](https://user-images.githubusercontent.com/12416002/161169364-d07214ba-27a8-4a22-a463-2b1826a0e9d6.png)

By the way, this seems to be a new issue. Should this one be closed?

**Comment by ssj71** [comment]
the `-dbg` flag disables midi so you should remove it once you are confident that the drumkit is picking up your hits properly. I'm still not confident that it is. If you just run `rbdrum2midi -v -rb1` what does the terminal output look like when you hit drums?

Also I'm not familiar with using amidi as a midi monitor instead of jack_midi_dump. If it's working for you no need to change, but if you aren't sure jsyk there are also alternatives like kmidimon and gmidimon.

**Comment by LaserTron** [comment]
It works more now! `amidi` outputs midi events. I plug the drums into Hydrogen and when I hit the pads the midi-in light lights up. Still no sound in Hydrogen. I plugged it into `qsynth` as well and again the midi lights up when I tap the pads, but no sound is produced. Let me know if you want more info.

I'll try tinkering with options in the programs.

**Comment by LaserTron** [comment]
Mini update, without the `-rb1` flag nothing works: hydrogen, amidi and qsynth don't report midi input.

**Comment by ssj71** [comment]
yes, from the data I see in your screenshot above your kit is the "rockband 1" style where there is no information about how hard you hit like there is in the "pro" kits. You'll always need the -rb1 flag to specify that the driver shouldn't watch for said data. There are quite a few possible reasons why you might not be getting sound out of H2 or qsynth. Make sure the drum is outputting on the correct midi channel, that's a common gotcha. :) Best of luck!

**Comment by ssj71** [comment]
Are we ok to close this then?
